packages/core/src/tracing/clsHooked/async_local_storage_context.js

```diff
@@ -0,0 +1,361 @@
+   * in this function will happen in that new context. The return value from that function (if any) is discarded.
+   * If you aren't 100% certain that the function never returns a value or that client code never relies on that value,
+   * use runAndReturn instead.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  run(fn, ctx) {
+    const context = ctx || this.createContext();
+
+    try {
+      storage.run(context, () => fn(context));
+      return context;
+    } finally {
+      this.exit(context);
+    }
+  }
+
+  /**
+   * Runs a function in a new CLS context and returns its return value. The context is left after the function
+   * terminates. Asynchronous work started in this function will happen in that new context.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runAndReturn(fn, ctx) {
+    let value;
+    this.run((/** @type {InstanaCLSContext} */ context) => {
+      value = fn(context);
+    }, ctx);
+    return value;
+  }
+
+  /**
+   * Runs a function which returns a promise in a new CLS context and returns said promise. The context is left
+   * as soon as the the promise resolves/is rejected. Asynchronous work started in this promise will happen in that new
+   * context.
+   *
+   * If the given function does not create a then-able, an error will be thrown.
+   *
+   * This function assumes that the returned promise is CLS-friendly or wrapped already.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runPromise(fn, ctx) {
+    const context = ctx || this.createContext();
+
+    // This updates the context in the promise, otherwise we return the context parent
+    // TODO: revisit this: do I need this enterWith?
+    storage.enterWith(context);
+    // this.enter(context);
+
+    return storage.run(context, () => {
+      const promise = fn(context);
+      if (!promise || !promise.then || !promise.catch) {
+        throw new Error('fn must return a promise.');
+      }
+
+      // TODO: just return promise
+      return promise
+        .then((/** @type {*} */ result) => {
+          // this.exit(context);
+          return result;
+        })
+        .catch((/** @type {*} */ err) => {
+          // this.exit(context);
+          throw err;
+        });
+    });
+  }
+
+  /**
+   * Runs a function (which might or might not return a promise) in a new CLS context. If the given function indeed
+   * returns a then-able, this behaves like runPromise. If not, this behaves like runAndReturn. In particular, no error
+   * is thrown if the given function does not return a promise.
+   *
+   * This function assumes that the returned promise (if any) is CLS-friendly or wrapped already.
+   * @param {Function} fn
+   * @param {InstanaCLSContext} ctx
+   */
+  runPromiseOrRunAndReturn(fn, ctx) {
+    let isPromise = false;
+    let valueOrPromise;
+    const context = ctx || this.createContext();
+    storage.enterWith(context);
```

Comment:
I need more investigation to prove why this is needed, as I really may have forgotten or didn't even really know, as a few changes involved trial and error, and this could be one of them.

All I can say for now is that if I remove this call to `enterWith`, there are several errors when I run the CI build. I can point that some of them happen in aws-lambda tests, but also in different places, like here: https://app.circleci.com/pipelines/github/instana/nodejs/2564/workflows/62a5d950-0bcf-49ee-841b-703d652f0f3a/jobs/11061/tests#failed-test-1

I will go back to this piece of the code and see if I can replace the `enterWith` call with a more satisfactory implementation.

Thanks